lib/recursive-methods.rb

```diff
@@ -1,49 +1,136 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError.new("Must be a positive integer.")
+  elsif n == 0 
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  if s.length == 0
+    return s
+  end
+  
+  i = (s.length - 1)
+  
+  if i == 0
+    return s
+  else
+    return s[i] + reverse(s[0..-2])
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def reverse_inplace(s, low = 0, high = s.length - 1)
+  
+  if low >= high
+    return s
+  else
+    temp = s[high]
+    s[high] = s[low]
+    s[low] = temp
+  end
+  
+  reverse_inplace(s, low +=1, high -= 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError.new("You can't have a negative number of bunnies, silly.")
+  elsif n == 0
+    return 0
+  elsif n == 1
+    return 2
+  else 2 + bunny(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
-def nested(s)
-    raise NotImplementedError, "Method not implemented"
+# Time complexity: O(n)
+# Space complexity: O(n)
+def nested(s, low = 0, high = s.length - 1)
+  status = true
+  
+  if s.length % 2 != 0
+    return false
+  elsif low >= high
+    return status
+  elsif s[low] == "(" && s[high] == ")"
+    status = true
+    nested(s, low += 1, high -= 1)
+  else
+    return false
+  end
+  
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def search(array, value)
```

Comment:
Because `array[1..-1]` makes a new array the time/space complexities are actually O(n<sup>2</sup>)